### DISCUSSION TITLE
Adding extra grayscale channel to multivariate view

### DISCUSSION THREAD

**Discussion by moon-chyldren** [Kitware/trame#650] [Q&A]
I've been playing with the [multivariate view](https://github.com/Kitware/multivariate-view) Trame project (found here https://github.com/Kitware/multivariate-view) and would like to know if it is possible to an extra channel that includes all of the channels as one grayscale model and then toggle the remaining channels as overlays within or on top of that one all inclusive grayscale channel?

Also, how can I add custom RGB values to each label/channel instead of having the app choose equidistant locations along the perimeter of the color picker wheel for each channel/label in my .HF file and assigning those colors to those labels/channels?

Can these both be achieved by using the `--opactity-channel` or other option in the arguments? I tried using the `--enable-preprocessing` option in the arguments and could not find any additional preprocessing control when I did.

**Comment by psavery** [comment]
> I've been playing with the [multivariate view](https://github.com/Kitware/multivariate-view) Trame project (found here https://github.com/Kitware/multivariate-view) and would like to know if it is possible to an extra channel that includes all of the channels as one grayscale model and then toggle the remaining channels as overlays within or on top of that one all inclusive grayscale channel?

It may be possible - something we would tackle in the Phase II if we get it. However, I'm unsure of how we would handle "blending" of the grayscale channel with the other channels within a given voxel. Currently, each voxel has an RGBA value that is computed based upon its position within the radial colormap. This extra grayscale channel would then be affecting this RGBA value somehow, and we'd need to do it in a way that's intuitive in visualization. It's a possibility! But something we would explore if we got the Phase II.

> Also, how can I add custom RGB values to each label/channel instead of having the app choose equidistant locations along the perimeter of the color picker wheel for each channel/label in my .HF file and assigning those colors to those labels/channels?

This is another possibility of something we could add if we got the Phase II :slightly_smiling_face:.

> Can these both be achieved by using the `--opactity-channel` or other option in the arguments? I tried using the `--enable-preprocessing` option in the arguments and could not find any additional preprocessing control when I did.

I think that `--enable-preprocessing` is set to `True` by default currently, so the option is only available to disable it.

For the `--opacity-channel`, you can indeed set one channel to correspond to opacity (and this channel gets excluded from the radial colormap). This channel would then be used directly to compute the alpha channel (`A`) in the `RGBA`. That additional opacity channel could be anything - it could potentially be an array derived from the other channels. Then, when you are visualizing the volume, more "opaque" voxels indicate a higher value in that opacity channel. It can be really helpful to, for example, decrease the visibility of voxels which are not high in some derived/computed property (and, conversely, increase the visibility of voxels that are high in that property).

**Reply by moon-chyldren** [reply]
Hopefully you get to Phase II! These features would be very useful with ouR data!

Responding to the following question and your response: what would be the best way to assign the custom RGB values programmatically? 

```
Also, how can I add custom RGB values to each label/channel instead of having the app choose equidistant locations 
along the perimeter of the color picker wheel for each channel/label in my .HF file and assigning those colors to 
those labels/channels?

          --This is another possibility of something we could add if we got the Phase II_ 🙂.
```

I've looked over the entire code base and it's not stated where in the code these RGB values should be assigned. I know it's a simple matter of hard coding those values into the code, it's just not immediately clear in which file and function I should plug those in. If you could point me to where  I could plug those in that would be extremely helpful.  🙂 

I'm putting on a short demo for some cohorts to demonstrate some of the capabilities of Trame in hopes of convincing them this is the solution we're looking for for examining the energy dispersive x-ray spectroscopy datasets we employ in our lab. 🤞 and This would be a good selling point for the custom colour labeling we use for material distributions.
